Documents the release process for Pyrsia (#1405)
pyrsia/pyrsia: docs/developers/release_process.md

# Making a Release for Pyrsia

Once the team decides to tag a release please use the following steps to ensure all parts of the service are updated.

## Before tagging the release in github

> - [] Make sure you update the version in `Cargo.toml`. Run the PR through the github actions to ensure nothing fails.

Reviewer: A white space is needed between brackets. It is not treated as a checkbox now. 
![image](https://user-images.githubusercontent.com/8938191/207529274-65394190-f6ea-462c-8ebd-3273eab93bf7.png)

> - [] Make sure you update the version in `Cargo.toml`. Run the PR through the github actions to ensure nothing fails.

Reviewer: Also we need to update GitHub Actions itself? According to the previous release - https://github.com/pyrsia/pyrsia/pull/1349/files

This is the PR that will be tagged with the version number that matches the one in the `Cargo.toml`

> - [] Run integration tests and ensure there are no failures - also ensure there no flakiness observed.

Reviewer 2: ```suggestion
- [] Run integration tests and ensure there are no failures - also ensure there is no flakiness observed.
```

- [] Record the results of the integration tests so that we can attach them to the release notes.
- [] Run manual confidence tests - this requires to run Pyrsia node as part of the network and exercising the new features manually. We may be able to automate parts of this but not completely since the network itself is complex to build a workflow on.

> - [] verify that the rust toolchain version is set to the version we would like to release this version. Since 0.2.2 this is captured in one place - at the top of `Cargo.toml`

Reviewer: It is better that all sentences start with a capital letter for consistency.

> - [] verify that the rust toolchain version is set to the version we would like to release this version. Since 0.2.2 this is captured in one place - at the top of `Cargo.toml`

Reviewer 2: ```suggestion
- [] Verify that the rust toolchain version is set to the version we would like to release. Since 0.2.2 this is captured in one place - at the top of `Cargo.toml`
```

- [] make sure apt repo and brew repo contain the correct latest release
- [] upload windows MSI to github release
- [] make sure the authorized nodes are updated
- [] run installer tests connecting to nightly
- [] update documentation to include latest released version

## Tag the release

- [] Commit the PR to the main branch

> - [] Once all the above steps are completed and verified to be success, tag the release in github. Github also helps you add the release notes - although they could use some context and explaning to make them easy to understand. Fill in that text to simplify the release notes and add the generated release notes in the release.

Reviewer 2: small typo
```suggestion
- [] Once all the above steps are completed and verified to be successful, tag the release in github. Github also helps you add the release notes - although they could use some context and explaining to make them easy to understand. Fill in that text to simplify the release notes and add the generated release notes in the release.
```

> ## After Tagging the release

Reviewer: Is this kind of blog post one of tasks to be listed here? https://pyrsia.io/blog/2022/11/30/pyrsia-0.2.1-released/

Author: indeed. Will add it to the list.

# Making a Release for Pyrsia

Once the team decides to tag a release please use the following steps to ensure all parts of the service are updated.

## Before tagging the release in github

- [ ] Make sure you update the version in `Cargo.toml`. Run the PR through the github actions to ensure nothing fails.
This is the PR that will be tagged with the version number that matches the one in the `Cargo.toml`
- [ ] Run integration tests and ensure there are no failures - also ensure there no flakiness observed.
- [ ] Record the results of the integration tests so that we can attach them to the release notes.
- [ ] Run manual confidence tests - this requires to run Pyrsia node as part of the network and exercising the new features manually. We may be able to automate parts of this but not completely since the network itself is complex to build a workflow on.
- [ ] Verify that the rust toolchain version is set to the version we would like to release this version. Since 0.2.2 this is captured in one place - at the top of `Cargo.toml`

> - [ ] Make sure apt repo and brew repo contain the correct latest release

Reviewer 3: this should go under Tag the release

Author: for `Tag the release` I kept only the steps that actually apply the tag. So moving items about running the tests as part of `After tagging`

> - [ ] Upload windows MSI to github release

Reviewer 3: this can only be done after tagging/building. move to 'Tag the release'

Author: created new headers for sections that are for testing and deployment.

> - [ ] Deploy the authorized nodes with this release

Reviewer 3: This can't be done before tagging. But I think this is a duplicate of `Deploy to production` which we already have under 'After tagging the release`

- [ ] Run installer tests connecting to nightly

> - [ ] Update documentation to include latest released version

Reviewer 3: Move to 'tag the release' or maybe even better to 'after tagging' - I think it's one the final steps

## Tag the release

> - [ ] Commit the PR to the main branch

Reviewer 3: We should discuss this: I don't think changing the version number e.g. from 0.2.2 to 0.2.3 should wait until we are ready to release 0.2.3 - in fact, I think it's the opposite, we should update to 0.2.3 right after we release 0.2.3.

If we agree on this, the very first bullet in this list is just a verification that the version in Cargo.toml is correct.
and the very last step should be to update the version to the next release version and create a PR for it

Author: that is correct we should update the version as soon as we have done the release. Although that is not the current state. So I will leave it that way and we will do it the way you are suggesting for 0.2.2 and update the document.

> - [ ] Update github actions with the new version number eg. <https://github.com/pyrsia/pyrsia/pull/1349/files>

Reviewer 3: this must be part of the very first bullet: Make sure you update the version in `Cargo.toml`
